blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/AbstractElasticIO.java

```diff
@@ -0,0 +1,321 @@
+    protected Histogram queryBatchHistogram = Metrics.histogram(getClass(), "Query Batch Size");
+
+    public static String METRICS_TOKENS_AGGREGATE = "metric_tokens";
+    public static String ELASTICSEARCH_INDEX_NAME_WRITE = Configuration.getInstance().getStringProperty(ElasticIOConfig.ELASTICSEARCH_INDEX_NAME_WRITE);
+    public static String ELASTICSEARCH_INDEX_NAME_READ = Configuration.getInstance().getStringProperty(ElasticIOConfig.ELASTICSEARCH_INDEX_NAME_READ);
+
+    public static String ENUMS_INDEX_NAME_WRITE = Configuration.getInstance().getStringProperty(ElasticIOConfig.ELASTICSEARCH_ENUMS_INDEX_NAME_WRITE);
+    public static String ENUMS_INDEX_NAME_READ = Configuration.getInstance().getStringProperty(ElasticIOConfig.ELASTICSEARCH_ENUMS_INDEX_NAME_READ);
+
+    private int MAX_RESULT_LIMIT = 100000;
+    public static final String REGEX_TOKEN_DELIMTER = "\\.";
+
+    //grabs chars until the next "." which is basically a token
+    private static final String REGEX_TO_GRAB_SINGLE_TOKEN = "[^.]*";
+
+
+    public List<SearchResult> search(String tenant, String query) throws Exception {
+        return search(tenant, Arrays.asList(query));
+    }
+
+
+    public List<SearchResult> search(String tenant, List<String> queries) throws Exception {
+        String[] indexes = getIndexesToSearch();
+
+        return searchESByIndexes(tenant, queries, indexes);
+    }
+
+    private List<SearchResult> searchESByIndexes(String tenant, List<String> queries, String[] indexes) {
+        List<SearchResult> results = new ArrayList<SearchResult>();
+        Timer.Context multiSearchCtx = searchTimer.time();
+        SearchResponse response;
+        try {
+            queryBatchHistogram.update(queries.size());
+            BoolQueryBuilder bqb = boolQuery();
+            QueryBuilder qb;
+
+            for (String query : queries) {
+                GlobPattern pattern = new GlobPattern(query);
+                if (!pattern.hasWildcard()) {
+                    qb = termQuery(ESFieldLabel.metric_name.name(), query);
+                } else {
+                    qb = regexpQuery(ESFieldLabel.metric_name.name(), pattern.compiled().toString());
+                }
+
+                bqb.should(boolQuery()
+                                .must(termQuery(ESFieldLabel.tenantId.toString(), tenant))
+                                .must(qb)
+                );
+            }
+
+            response = client.prepareSearch(indexes)
+                    .setRouting(tenant)
+                    .setSize(MAX_RESULT_LIMIT)
+                    .setVersion(true)
+                    .setQuery(bqb)
+                    .execute()
+                    .actionGet();
+        } finally {
+            multiSearchCtx.stop();
+        }
+
+
+        for (SearchHit hit : response.getHits().getHits()) {
+            SearchResult result = convertHitToMetricDiscoveryResult(hit);
+            results.add(result);
+        }
+        return dedupResults(results);
```

Comment:
This is fine, but to me it's strange that you build a result list above and then trim the dups. Why not just build non-duped list in the first place? It's possible I'm missing a glaring fact. :-)